at_cmds-1/aurpd.tproj/aurpd.8
.TH aurpd 8
.SH NAME
.B aurpd 
- Start up the AURP routing function

.SH SYNOPSIS
.B aurpd 
[-dq] [-l 
.I logfile
] [
.I config_file
]

.SH FLAGS
.TP 8
-d
terminate an existing aurp daemon
.TP 8
-q 
"quiet" - most error and other messages are suppressed
.TP 8
-l 
log to the specified file (normally stdout)
.TP 8
.I config_file
The file 
.B aurpd 
uses to configure itself
(cf. aurp_tunnel.cfg(5)).  The default, if a name is not provided, is
"/etc/aurp_tunnel.cfg".

.SH DESCRIPTION

The 
.B aurpd 
daemon initializes and configures kernel support for AURP,
Apple's Update Routing Protocol.  The configuration file is read and
parsed, and the kernel's AURP tunneling support is set up.

Currently, the following AURP functionality is supported:
.TP 8
	- tunneling through UDP
.TP 8
	- explicit network number hiding
.TP 8
	- explicit network number advertising.

.in 7
The configuration file specifies the endpoints, local and remote, of
the tunnels to be configured.  The tunnel will provide access to all
AppleTalk network numbers reachable from the local endpoint, unless
the 'hide' or 'export' keywords are specified in the file.  See
aurp_tunnel.cfg(5) for more details.

The AURP daemon will remain running until another instance of the
command is run with the '-d' flag to terminate the session.  When run
with the '-d' flag, the 'aurpd' command will invoke a kernel operation
that will shut down the tunnel and eventually cause the existing
'aurpd' process to exit after it receives a message from the stack,
indicating that shutdown is complete.

.SH RETURN VALUES
.TP 8
	 0 - Clean exit
.TP 8
	-1 - Not super-user (must be root)
.TP 8
	-3 - Problem parsing the config file
.TP 8
	-4 - Can't fork
.SH BUGS
The tunnel mechanism doesn't lend itself to use on a multihomed host
that doesn't enable forwarding for IP packets.  The '-v' switch,
although specified in the "usage" printout, will only cause aurpd to
complain about an unknown flag.

.SH SEE ALSO
appletalk(8), aurp_tunnel.cfg(5)


objc-1/ChangeLog
Changes from Ali to all the string class files. (12/23/92)
Changes from Derek to deal with name conflicts.

Cleaned up some warnings (7/23/91):

	* Makefile (COMMONFLAGS):  Changed -Wimplicit to -Wall.
	Let's make this compile	without warnings!

	* List.m, Storage.m (read:, write:):  Changed to return self.

	* <various files>:  Removed all unused local variables.

Added support for string objects (7/23/91):

	* NXString.m, NXCharacterSet.m, unichar.h, StringTest.m:  New files.

	* Makefile (LOCAL_CLASSES):  Added NXString.m and NXCharacterSet.m

	* Makefile (LOCAL_INCFILES):  Added unichar.h

	* objc-runtime.m (_objc_fixup_string_object_class):  New routine.
	If NXConstantString is present in class hash table, copy it to
	_NXConstantString_class and insert this address in class hashtable.

	* objc-runtime.m (_objcInit):  Call new routine after classes have
	been fixed up (in _objc_map_selectors).

	* objc-globaldata.m (_NXConstantString_class):  New global variable.
	Reduce pad by 36 bytes to compensate.

	* spec_sys:  Added slots for NXCompareCharacters(), NXEmptyString(),
	and NXFindCharacters().  Added files NXString.o and NXCharacterSet.o
	These changes will need to be integrated into libsys's copy.

Changed -[Object copy] (7/24/91):

	* Object.m (copy):  Chnaged to return [self copyFromZone: [self zone]]
	rather than using _copy.  The function object_copy() is no longer used
	by the runtime system, and could probably be removed.

	* List.[hm] (copy):  Deleted redundant method.

	* Storage.[hm] (copy):  Deleted redundant method.

	* HashTable.[hm] (copy):  Deleted redundant method.

	* NXString.[hm] (copy):  Deleted redundant method.

	* NXCharacterSet.[hm] (copy):  Deleted redundant method.

Changed method names in Storage (7/24/91):

	* Storage.[hm] (replaceElementAt:with:)  Changed from replace:at:
	and added 2.0 compatibility method in implementation only.

	* Storage.[hm] (insertElement:at:)  Changed from insert:at:
	and added 2.0 compatibility method in implementation only.

	* Storage.[hm] (removeElementAt:)  Changed from removeAt:
	and added 2.0 compatibility method in implementation only.

Cleaned up Makefile slightly (7/24/91):

	* Makefile (BY_PRODUCTS):  Added objcopt so that make clean removes
	all files not installed by make installsrc.

Fixed bug #14923 in HashTable (7/24/91):

	* HashTable.m (nextState:key:value:):  Set key and value to NULL
	when returning NO.

Fixed bug #14309 in Storage (7/24/91):

	* Storage.m (setNumSlots:):  Changed incorrect test
	(numSlots > maxElements) to (maxElements > old).

	* Storage.m (setNumSlots:):  Changed to use a single realloc.

Fixed bug #5687 in hashtable (7/24/91):

	* hashtable.h (NXHashInsertIfAbsent):  Improved comment.

Added appendList: method (dmoore 7/24/91)

	* List.m, List.h added method.
	
Added maptable functionality (dmoore 7/24/91)

	* maptable.c maptable.m

Removed typedef for id when compiling Objective-C (7/26/91):

	* objc.h (id):  Do not define typedef for id when compiling
	Objective-C (when __OBJC__ is predefined).
	
Made Objective C check for duplicate classes and use the one in the application
	(dmoore 7/27/91)
	
	* objc-runtime.m:  Added following functions:
		_objc_inform:	Just prints message for now.
		_resolveClassConflict: Makes sure that old class goes back in
			if it's from the app.
		_nameForHeader:	Returns the name of an image.
		_getExecHeader:	Picks out the app header.
		_headerForClass: Finds the mach_header containing a class.
		
		Modified _objc_get_classes to check for duplicates.
		
Added maptable prototypes to objc-globaltext.m (dmoore 7/27/91)

Modified spec_sys to include NXString and maptable stuff.

snaroff (7/31/91):

Integrate blaines kernel mods in a rational way.

Renamed "_objc_internal_error" to "_objc_inform"...removed "_objc_inform",
they were nearly identical (snaroff 7/31/91)

Create file "objc-errors.m"...it includes _objc_fatal, _objc_inform,
__objc_error, and _objc_error. MUST ADD _objc_inform and NXLogError to list 
of private externs!!!!!

Call _objc_fatal() instead of perror() in objc-sel.m...callings perror in these situations is bogus.

Have hashtable use NXLogError instead of fprintf(). 

Add field to objc_class for protocols...add protocol templates to objc-class.h

Installed new NXString files from ali (8/3/91):

	* NXString.[hm]:  Most methods are now statically typed.
	
	* NXStringTest.m  Changes to match.

Changes for clean compilation with new compiler (8/3/91):

	* Makefile (objcopt):  Use -ObjC flag with objc-edit.c

	* objcedit.c (round):  Made declaration global as well as definition.

	* objcedit.c (map_input):  Cast third arg to map_fd().

	* objcopt.c (get_objc):  Cast to (SEL *) two places.

	* objc-sel.m (_sel_writeHashTable):  Cast to (POHASH) two places.

	* objc-class.m (cache_flush, _cache_expand, _cache_flush):
	Made iteration variable unsigned.

	* hashtable.m (CopyIntoReadOnly):  Made iteration variable unsigned.

	* typedstream.m (InternalWriteObject):  Cleaned up printf format
	specifiers to match args.

	* typedstreamprivate.h:  Changed all function declarations to static.

	* objc-utils.m:  Cleaned up printf format specifiers to match args.

	* NXString.m (getBytes:range:, getCharacters:range:, NXFindCharacters,
	find:range:flags:usingTable:, _debugPrint):
	Made iteration variable unsigned.

	* NXCharacterSet.m (init, addCharacters:length:,
	removeCharacters:length:):  Made iteration variable unsigned.

Improvements to NXConstantString's (8/12/91):

	* NXString.[hm] (new):  Add version which raises exception.

	* NXString.[hm] (copyFromZone:):  Return an NXReadOnlyString copy
	instead of self, since our code might be unloaded.

	* NXString.[hm] (write:):  Write an NXReadOnlyString copy.

	* NXString.[hm] (finishUnarchiving):  Removed.

	* NXString.[hm] (free):  This now raises an error, since copyFromZone:
	no longer returns self.  You never allocate NXConstantString's, so you
	should never free them.


at_cmds-1/aurpd.tproj/aurpd.c
/*
 * Copyright (c) 1999 Apple Computer, Inc. All rights reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 * 
 * "Portions Copyright (c) 1999 Apple Computer, Inc.  All Rights
 * Reserved.  This file contains Original Code and/or Modifications of
 * Original Code as defined in and that are subject to the Apple Public
 * Source License Version 1.0 (the 'License').  You may not use this file
 * except in compliance with the License.  Please obtain a copy of the
 * License at http://www.apple.com/publicsource and read it before using
 * this file.
 * 
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE OR NON-INFRINGEMENT.  Please see the
 * License for the specific language governing rights and limitations
 * under the License."
 * 
 * @APPLE_LICENSE_HEADER_END@
 */
/*
 *	Copyright (c) 1996 Apple Computer, Inc.
 *
 *	The information contained herein is subject to change without
 *	notice and  should not be  construed as a commitment by Apple
 *	Computer, Inc. Apple Computer, Inc. assumes no responsibility
 *	for any errors that may appear.
 *
 *	Confidential and Proprietary to Apple Computer, Inc.
 *	Created April 8, 1996 by Tuyen Nguyen
 *	Modified, May, 1996, to move routing/tunneling into kernel.
 *		aurpd now functions as user process to configure/update
 *		kernel process.
 *	Modified, March 17, 1997 by Tuyen Nguyen for Rhapsody.
 */
#ifndef lint
static char sccsid[] = "@(#)aurpd.c; Copyright 1996, Apple Computer, Inc.";
#endif  /* lint */

/*
 * Exit codes:
 *	0 - AOK
 *	-1 - Not Super-User
 *	-3 - Bad config file
 *	-4 - Fork failed
 */
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <h/sysglue.h>
#include <at/appletalk.h>
#include <at/aurp.h>
#ifdef _AIX
#include <locale.h>
#include <atsbin_msg.h>
#endif

/* */
#define AURP_CFGFILENAME  "/etc/aurp_tunnel.cfg"

/*#define MSGSTR(num,str)   catgets(catd, MS_SBIN, num,str)*/
#define MSGSTR(num,str)		str

int  AURPgetcfg();
void AURPdown();
#ifdef _AIX
extern char *sys_errlist [];
nl_catd catd;
#endif
char *usage = "Usage: %s [-dqv] [-l log] [config_file]\n";
char *progname;
#ifdef DEBUG
char *opt_string = "Ddl:qv";
#else
char *opt_string = "dl:qv";
#endif
FILE *logfp;
char *logfn;
short udp_port;
unsigned char opt_error = 1;	/* Print helpful message on error */
unsigned char dst_addr_cnt;
unsigned char net_export;
unsigned char net_hide;
unsigned char net_access_cnt;
unsigned char debug;
unsigned short net_access[AURP_MAXNETACCESS];
char dst_active[256];
long dst_addr[256];
long src_addr;

main(argc, argv)
	int	argc;
	char *argv[];
{	int rc;
	char *filename = AURP_CFGFILENAME;
	register int errflag, c;
	extern int optind;
	extern int optopt;
	extern int opterr;
	extern char *optarg;
	extern void AURP();
#ifdef _AIX
	(void) setlocale(LC_ALL, "");
	catd = catopen(MF_ATSBIN, NL_CAT_LOCALE);
#endif

	/*
	 * find out what this program is called and other miscellaneous stuff
	 */
	progname = strrchr(argv[0], '/') ? strrchr(argv[0], '/') + 1 : argv[0];

	errflag = 0;
	while ((c = getopt(argc, argv, opt_string)) != EOF)
		switch(c)
		{
#ifdef DEBUG
			case 'D':
				debug++;
				break;
#endif
			case 'd':
				AURPdown();
#ifdef _AIX
				catclose(catd);
#endif
				exit(0);
			case 'l':
				logfn = optarg;
				break;
			case 'q':
				opt_error = 0;
				break;
			case '?':
			default:
				errflag++;
				break;
		}

	if (errflag)
	{	fprintf(stderr, MSGSTR(M_USAGE, usage), progname);
		return -1;
	}

	/*
	 * Make sure that program is invoked by a super-user, before we
	 *  do anything substantive.
	 */
	if (getuid()) {
		fprintf(stderr, MSGSTR(M_MUST_BE_SU,
			"%s: Permission denied; must be super-user.\n"), progname);
#ifdef _AIX
		catclose(catd);
#endif
		exit(-1);
	}

	if (optind < argc)
		filename = argv[optind];

	if (AURPgetcfg(filename)) {
#ifdef _AIX
		catclose(catd);
#endif
		exit(-3);
	}

	/* run the aurp daemon in the background, disassociated from tty */
#ifdef DEBUG
	/* if we're not debugging and the user wants to */
	if (!debug)
	{
#endif
	if ((rc = fork()) == -1) {
		perror("fork");
#ifdef _AIX
		catclose(catd);
#endif
		exit(-4);
	}
	if (rc) {		/* In parent */
#ifdef _AIX
		catclose(catd);
#endif
		exit(0);
	}
#ifdef DEBUG
	}
#endif

	setpgid(0, 0);
	AURP();
#ifdef _AIX
	catclose(catd);
#endif
	exit(0);
}

void AURP()
{
	int at_descr, ic_cmd, ic_len;
	ioccmd_t ioc;

	/* create the AURP Appletalk socket */
	if ((at_descr = at_open_dev(ATPROTO_AURP)) == -1) {
		perror("at_open_dev");
		return;
	}

	/* configure local/remote addresses */
	if (dst_addr_cnt) {
		ic_cmd = AUC_NETLIST;
		ic_len = (dst_addr_cnt+1) * sizeof(unsigned long);
		if (at_send_to_dev(at_descr, ic_cmd, dst_addr, &ic_len) == -1) {
			perror("NETLIST");
			close(at_descr);
			return;
		}
	}
	/* configure network hiding or exporting */
	if (net_access_cnt) {
		ic_cmd = net_export ? AUC_EXPNET : AUC_HIDENET;
		ic_len = net_access_cnt * sizeof(unsigned short);
		if (at_send_to_dev(at_descr, ic_cmd, net_access, &ic_len) == -1) {
			perror("HIDE/EXPORT");
			close(at_descr);
			return;
		}
	}

	if (udp_port != AURP_SOCKNUM)
	{	ic_cmd = AUC_UDPPORT;
		ic_len = sizeof(udp_port);
		if (at_send_to_dev(at_descr, ic_cmd, &udp_port, &ic_len) == -1) {
			perror("UDPPORT");
			close(at_descr);
			return;
		}
	}

	/* configure the tunnel - this starts the ball rolling */
	ic_cmd = AUC_CFGTNL;
	ic_len = sizeof(dst_addr_cnt);
	if (at_send_to_dev(at_descr, ic_cmd, &dst_addr_cnt, &ic_len) == -1) {
		perror("AURP CONFIG");
		close(at_descr);
		return;
	}

	logfp = stdout;
	if (logfn) {
		/* log data to the specified file */
		if ((logfp = fopen(logfn, "w")) == 0) {
			perror(logfn);
			close(at_descr);
			return;
		}
	}

	/* wait for shutdown */
	errno = aurpd(at_descr);

	close(at_descr);
}

int AURPgetcfg(filename)
	char *filename;
{
	FILE *fp;
	int line, len;
	char cfgbuf[128];

	/* open the cfg file */
	errno = 0;
	if ((fp = fopen(filename, "r")) == 0) {
		perror(filename);
		return -1;
	}

	/* get the cfg info */
	udp_port = -1;
	dst_addr[0] = -1;
	for (line=1; !feof(fp); line++) {
		register char *p, *q;

		/* fgets() NUL-terminates our string, but leaves the NL */
		if (fgets(cfgbuf, sizeof(cfgbuf), fp) == 0)
		{	if (errno)
				perror(filename);
			break;
		}
		if (cfgbuf[0] == '#')
			continue;
		if ((p = strtok(cfgbuf, "\t \n")) == NULL)	/* Keyword */
			continue;
		q = strtok(NULL, "\t \n");	/* value */

		if (strpfx(p, "port") == 0)
		{	if (udp_port == -1)
				udp_port = (short)atoi(q);
		} else if (strpfx(p, "local") == 0)
		{	if (dst_addr[0] == -1)
				dst_addr[0] = inet_addr(q);
		} else if (strpfx(p, "remote") == 0)
		{	dst_addr[++dst_addr_cnt] = inet_addr(q);
		}
		else if (strpfx(p, "hide") == 0)
		{	if (net_export) {
				fprintf(stderr,
					MSGSTR(M_HIDE, "%s: Hiding/Export conflict\n"),
					progname);
				return -1;
			}
			net_hide = 1;
			net_access[net_access_cnt++] =
				(unsigned short)atoi(q);
		} else if (strpfx(p, "export") == 0)
		{	if (net_hide) {
				fprintf(stderr,
					MSGSTR(M_EXP, "%s: Export/Hiding conflict\n"),
					progname);
				return -1;
			}
			net_export = 1;
			net_access[net_access_cnt++] =
				(unsigned short)atoi(q);
		} else
		{	if (opt_error)
				fprintf(stderr,
					MSGSTR(M_KEYWORD,
					       "%s: unknown keyword (`%s\'), line %d\n"),
					filename, p, line);
			continue;
		}
	}

	if (dst_addr_cnt == 0) {
		if (opt_error)
			fprintf(stderr,
				MSGSTR(M_REMOTE,
				       "%s: Invalid format (%s): no `remote\' keyword\n"),
				progname, filename);
		return -1;
	}

	if (src_addr == -1) {
		fprintf(stderr,
			MSGSTR(M_LOCAL,
			       "%s: Invalid format (%s): no `local\' keyword\n"),
			progname, filename);
		return(-1);
	}

	if (udp_port == -1)
		udp_port = AURP_SOCKNUM;

	return 0;
}

void AURPdown()
{
	int at_descr, ic_cmd;
	ioccmd_t ioc;

	if ((at_descr = at_open_dev(ATPROTO_AURP)) == -1)
		perror("at_open_dev");
	else {
		ic_cmd = AUC_SHTDOWN;
		if (at_send_to_dev(at_descr, ic_cmd, 0, 0) == -1)
			perror("SHUTDOWN");
		close(at_descr);
	}
}

/* see if pat is a prefix of str */
strpfx(pat, str)
register char *pat, *str;
{
        if (pat == NULL || str == NULL)
                return(0);
        if (*pat == '\0' || *str == '\0')
                return(0);
        while (*pat && *str && *pat == *str)
        {       pat++;
                str++;
        }
        /* If *pat is nul, pat is a prefix; otherwise, not */
        return(*pat);
}
